File: Python3/0289_game_of_life.py
# https://leetcode.cn/problems/game-of-life/

class Solution:
    def gameOfLife(self, board):
        """
        Do not return anything, modify board in-place instead.
        """

        """
        Inputs:
            board: list[list[int]]
        Outputs:
            None
        """
        rows = len(board)
        cols = len(board[0])
        for row in range(rows):
            for col in range(cols):
                # Check one cell.
                cnt = 0
                for i in range(-1, 2):
                    r = row + i
                    for j in range(-1, 2):
                        c = col + j
                        if (r < 0) or (r >= rows) or (c < 0) or (c >= cols):
                            continue
                        if (r == row) and (c == col):
                            continue
                        if (board[r][c] == 1) or (board[r][c] == 3):
                            cnt = cnt + 1
                # 2: 0 -> 1
                # 3: 1 -> 0
                if board[row][col] == 1:
                    if cnt < 2 or cnt > 3:
                        board[row][col] = 3
                else:
                    if cnt == 3:
                        board[row][col] = 2

        for row in range(rows):
            for col in range(cols):
                if board[row][col] == 2:
                    board[row][col] = 1
                elif board[row][col] == 3:
                    board[row][col] = 0

File: Cpp/0289_game_of_life.cpp
// https://leetcode.cn/problems/game-of-life/

class Solution {
public:
    void gameOfLife(vector<vector<int>>& board) {
        int rows = board.size(), cols = board[0].size();
        for (int row = 0; row < rows; ++row) {
            for (int col = 0; col < cols; ++col) {
                // Check one cell.
                int cnt = 0;
                for (int i = -1; i < 2; ++i) {
                    for (int j = -1; j < 2; ++j) {
                        int r = row + i, c = col + j;
                        if ((r < 0) || (r >= rows) || (c < 0) || (c >= cols)) {
                            continue;
                        }
                        if ((r == row) && (c == col)) {
                            continue;
                        }
                        // 2: 0 -> 1
                        // 3: 1 -> 0
                        if ((board[r][c] == 1) || (board[r][c] == 3)) {
                            cnt = cnt + 1;
                        }
                    }
                }
                // Update rules.
                if (board[row][col] == 1) {
                    if ((cnt < 2) || (cnt > 3)) {
                        board[row][col] = 3;
                    }
                }
                else {
                    if (cnt == 3) {
                        board[row][col] = 2;
                    }
                }
            }
        }

        for (int row = 0; row < rows; ++row) {
            for (int col = 0; col < cols; ++col) {
                if (board[row][col] == 3) {
                    board[row][col] = 0;
                }
                else if (board[row][col] == 2) {
                    board[row][col] = 1;
                }
            }
        }
    }
};
